src/inmanta/agent/handler.py

```diff
@@ -434,16 +434,74 @@ class ResourceHandler(object):
 
     def __init__(self, agent: "inmanta.agent.agent.AgentInstance", io: Optional["IOBase"] = None) -> None:
         self._agent = agent
-
         if io is None:
             raise Exception("Unsupported: no resource mgmt in RH")
         else:
             self._io = io
-
         self._client: Optional[protocol.SessionClient] = None
         # explicit ioloop reference, as we don't want the ioloop for the current thread, but the one for the agent
         self._ioloop = agent.process._io_loop
 
+    # Interface
+
+    def close(self) -> None:
+        pass
+
+    @abstractmethod
+    def deploy(
+        self,
+        ctx: HandlerContext,
+        resource: resources.Resource,
+        requires: Dict[ResourceIdStr, ResourceState],
+    ) -> None:
+        """
+        Main entrypoint of the handler that will be called by the agent to deploy a resource on the server.
+        This method is always called by the agent, even when one of the requires of the given resource
+        failed to deploy.
+
+        :param ctx: Context object to report changes and logs to the agent and server.
+        :param resource: The resource to deploy
+        :param requires: A dictionary mapping the resource id of each dependency of the given resource to its resource state.
+        """
+        pass
+
+    @abstractmethod
+    def execute(self, ctx: HandlerContext, resource: resources.Resource, dry_run: bool = False) -> None:
```

Comment:
What is the difference between the `deploy()` and the `execute()` method in the context of a `DiscoveryHandler`?

Reply:
dryrun still uses execute instead of deploy

Reply:
Ok. What would be the main entrypoint for the `DiscoveryHandler`? Is it `deploy()` or `execute()`. Because I would find `deploy()` a strange name. It doesn't really deploy anything in the end.

Reply:
I agree, I'm trying to come up with some clearer phrasing

Reply:
I added a suggested change to the `deploy` docstring to try and clear this up. What do you think? Anything missing or still unclear there? After we've reached consensus there I think we need to clear up this docstring in the same vein.